core/utils/log.cpp

```diff
@@ -65,47 +67,96 @@ NS_LOCAL
 
 MSVC_ONLY(__pragma(warning(push)))
 MSVC_ONLY(__pragma(warning(disable: 4996))) // the compiler encountered a deprecated declaration
-static FILE* dev_null() {
-  return nullptr;
-}
+
 MSVC_ONLY(__pragma(warning(pop)))
 
+void noop_log_appender(void*, char const*, char const*,
+                       int, irs::logger::level_t,
+                       char const*, size_t) {}
+
+
+void fd_log_appender(void* context, char const* function, char const* file, int line,
+    irs::logger::level_t level, char const* message,
+    size_t) {
+  FILE* fd = reinterpret_cast<FILE*>(context);
+  if (fd != nullptr) {
+    char const* prefix = "UNKNOWN";
+    switch (level) {
+      case irs::logger::IRL_FATAL: prefix = "FATAL"; break;
+      case irs::logger::IRL_ERROR: prefix = "ERROR"; break;
+      case irs::logger::IRL_WARN: prefix = "WARN"; break;
+      case irs::logger::IRL_INFO: prefix = "INFO"; break;
+      case irs::logger::IRL_DEBUG: prefix = "DEBUG"; break;
+      case irs::logger::IRL_TRACE: prefix = "TRACE"; break;
+    }
+    std::fprintf(fd, "%s: %s:%d %s\n", prefix, (file ? file : "<no file provided>"), line, (message ? message : "<no message provided>"));
+    if (irs::logger::IRL_FATAL == level) {
+      std::fflush(fd);
+    }
+  }
+}
+
 class logger_ctx: public iresearch::singleton<logger_ctx> {
  public:
   logger_ctx()
     : singleton(),
       stack_trace_level_(irs::logger::level_t::IRL_DEBUG) {
     // set everything up to and including INFO to stderr
     for (size_t i = 0, last = iresearch::logger::IRL_INFO; i <= last; ++i) {
-      out_[i].file_ = stderr;
+      output(static_cast<iresearch::logger::level_t>(i), fd_log_appender, stderr);
     }
   }
 
-  bool enabled(iresearch::logger::level_t level) const { return dev_null() != out_[level].file_; }
-  FILE* file(iresearch::logger::level_t level) const { return out_[level].file_; }
+  bool enabled(iresearch::logger::level_t level) const { return noop_log_appender != out_[level].appender_.load(std::memory_order_relaxed); }
 
-  logger_ctx& output(iresearch::logger::level_t level, FILE* out) {
-    out_[level].file_ = out ? out : dev_null();
+  logger_ctx& output(iresearch::logger::level_t level, iresearch::logger::log_appender_callback_t appender, void* context) {
+    static std::mutex mutex;
+    SCOPED_LOCK(mutex); // only log level parallel updates are not thread safe. Using log contexts is safe even while concurrent update is ongoing
+    if (appender != nullptr) {
+      auto& ctx = out_[level];
+      ctx.appender_.store(noop_log_appender, std::memory_order_relaxed); // to play safe - as noop never uses context and noop log never breaks
+      ctx.appender_context_.store(context, std::memory_order_release);
+      ctx.appender_.store(appender, std::memory_order_relaxed);
```

Comment:
I suggest to utilize 16-bytes atomic instead, e.g.
```
struct alignas(IRESEARCH_CMPXCHG16B_ALIGNMENT) appender_context {
  void* context{nullptr};
  log_appender_callback_t appender{noop_log_appender};
};
```
see `object_pool.hpp` for more info